### PR TITLE
Implement utmp audit

### PR DESCRIPTION
**Pull request by stlaz**
Implements utmp audit required by policykit.

This commit also flattens the pidtree for the DM by starting the
display environment directly in the first fork which already should
have the environment ready for this purpose.

Fixes https://github.com/cylgom/ly/issues/121

### CODE REVIEW

File: src/login.c [hunk 1]
```diff
@@ -268,6 +269,38 @@ void env_xdg(const char* tty_id, const enum display_server display_server)
...
+	struct utmp *entry,
+	char *username,
+	pid_t display_pid
+) {
+	entry->ut_type = USER_PROCESS;
+	entry->ut_pid = display_pid;
+	strcpy(entry->ut_line, ttyname(STDIN_FILENO) + strlen("/dev/"));
+
+    /* only correct for ptys named /dev/tty[pqr][0-9a-z] */
+    strcpy(entry->ut_id, ttyname(STDIN_FILENO) + strlen("/dev/tty"));
```
Comment (roosemberth) [inline src/login.c]: Hi, could you please fix the indentation in these lines?

Comment (stlaz) [reply src/login.c]: oops, I thought I fixed them all. It should now be properly indented.

File: src/login.c [hunk 2]
```diff
@@ -609,17 +618,28 @@ void auth(
...
+	struct utmp entry;
+	add_utmp_entry(&entry, pwd->pw_name, pid);
+
 	// wait for the session to stop
 	int status;
 	waitpid(pid, &status, 0);
+	remove_utmp_entry(&entry);
+
+	reset_terminal(pwd);
+	if (desktop->display_server[desktop->cur] == DS_XORG && xcb != NULL)
```
Comment (stlaz) [inline src/login.c]: I just realized this will need additional check for DS_XINITRC, I’ll fix that when I get to my laptop

Comment (stlaz) [reply src/login.c]: FIXED by initializing the `xcb` variable (also fixed a possible segfault), the X session still hangs as described in #121 though, gotta investigate why

**Comment by cylgom** [comment]
Hi, thanks a lot! I was working on the same thing but you're far more advanced than I so we should probably merge your PR instead.

I also have the issues you and @evgenux described in #121 so I tried to investigate a little bit:
 - Xorg stays open after the desktop environment is closed
 - If you kill it, ly will hang
 - With the proprietary nvidia driver, X11 desktop environments refuse to start

Maybe my understanding is wrong, but for me:
 - We should keep that extra process to launch Xorg
 - We can't rely on the existing environment (the linux `login` from `login-utils` handles everything itself, so maybe we should, too)
 - None of this affected Sway, so the issue definitely comes from the X11 side

I am confident we will get this working though, and thanks again for contributing :)

**Comment by stlaz** [comment]
Hi,
Sorry for not replying earlier, that's some good research you did.

I too noticed that Xorg stays open, I also noticed that with this patch, on logout, the pstree of the created fork gets transferred to the `init` process upon logout, which may indicate that our parent process exits, I wonder if that's caused by the terminal reset that we run or some unhandled signal sent from the child.

I think killing Xorg might be the way for now, if ly hangs, then I'll need to figure out why that is, as that should not happen.

I wonder about the proprietary driver though. I know `sway` does not like them on purpose, but I didn't know it may cause trouble to X11 desktop environments, too.

I would still like to flatten the processes if there's a way to do that.

File: src/login.c [hunk 3]
```diff
@@ -526,6 +531,9 @@ void auth(
 	tb_present();
 	tb_shutdown();
 
+	// we need xcb connection to clean up after ourselves if running xorg
+	xcb_connection_t* xcb = NULL;
```
Comment (stlaz) [inline src/login.c]: This was silly, we will need shared memory here

**Comment by stlaz** [comment]
@cylgom you were right, dealing with X11 is a bit too painful, I removed the process tree flattening for that.

Please review the changes again, see if they are ok

**Comment by cylgom** [comment]
Wonderful! Thank you very much :)

**Pull request merged**